packages/astro/src/runtime/client/dev-toolbar/apps/audit/index.ts

```diff
@@ -69,8 +70,46 @@ export default {
 
 		await lint();
 
-		document.addEventListener('astro:after-swap', async () => lint());
-		document.addEventListener('astro:page-load', async () => refreshLintPositions);
+		let mutationDebounce: ReturnType<typeof setTimeout>;
+		const observer = new MutationObserver(() => {
+			if (mutationDebounce) {
+				clearTimeout(mutationDebounce);
+			}
+			mutationDebounce = setTimeout(() => {
+				settings.logger.verboseLog('Rerunning audit lints because the DOM has been updated.');
+
+				if ('requestIdleCallback' in window) {
+					window.requestIdleCallback(
+						async () => {
+							lint();
+						},
+						{ timeout: 300 }
+					);
+				} else {
+					// Fallback for old versions of Safari
+					setTimeout(() => {
+						lint();
+					}, 150);
+				}
+			}, 250);
+		});
+
+		setupObserver();
+
+		document.addEventListener('astro:before-preparation', () => {
+			observer.disconnect();
+		});
+		document.addEventListener('astro:after-swap', async () => {
+			lint();
+		});
+		document.addEventListener('astro:page-load', async () => {
+			refreshLintPositions();
+
+			// HACK: View transitions add a route announcer after this event, so we need to wait for it to be added
+			setTimeout(() => {
+				setupObserver();
```

Comment:
Query: we call `setupObserver` twice (here and at line 97), does that mean we end up observing `document.body` twice?

Reply:
No, because we disconnect the observer on page preparation. (line 100) Additionally, it's not possible for a single observer to watch the same element twice

Per MDN:
> If you call observe() on a node that's already being observed by the same MutationObserver, all existing observers are automatically removed from all targets being observed before the new observer is activated.

The final `expect` in the `handle mutations properly during view transitions` test also test this, by checking that there's only one instance of "Rerunning lints..."